lib/uart.cpp

```diff
@@ -98,47 +122,59 @@ uint8_t Uart::read(Uart::Address register_address) {
                 }
                 return data;
             }
-            break;
         }
         case Uart::Address::IER: {//IER or DLM
             if (this->dlab()) {//DLM
                 return this->regs.m_dlm;
             } else {//IER
-                assert(false && "TODO");//TODO
+                return this->regs.m_ier;
             }
-            break;
         }
         case Uart::Address::ISR: {//NOTE: Never FCR since that isn't readable
-            assert(false && "TODO");//TODO
-            break;
+            uint8_t isr = 0;
+            isr |= (this->regs.m_fcr & 0b1) ? (0b11 << 6) : 0;//Indicate the FIFOs are enabled if they infact are
+
+            //If there is a character available to read, and the Received Data Ready interrupt is enabled
+            if ((receive_queue.size() > 0) && (this->regs.m_ier & 0b1)) {
+                isr |= 0b0100;
+            } else if (this->regs.m_ier & 0b10) {//If the THR empty interrupt is enabled (since we are always ready to transmit)
+                isr |= 0b0010;
+            } else {//No enabled interrupts pending
+                isr |= 0b0001;
+            }//We don't need to support any other interrupt types
+
+            return isr;
         }
         case Uart::Address::LCR: {
-            assert(false && "TODO");//TODO
-            break;
+            return this->regs.m_lcr;
         }
         case Uart::Address::MCR: {
-            assert(false && "TODO");//TODO
-            break;
+            return this->regs.m_mcr;
         }
         case Uart::Address::LSR: {
-            //Clear the Data Ready bit in the LSR if the queue is empty
-            if (receive_queue.size() == 0) {
-                constexpr uint32_t LSR_DATA_READY_POS = 0U;
-                this->regs.m_lsr &= ~(1U << LSR_DATA_READY_POS);
-                
-                //TODO do we need to clear the interrupt pending bit too?
+            if (this->dlab()) {
+                irvelog(0, "Software tried to read from the UART's PSD register, which is write only!");
+                return this->regs.m_psd;
             }
-
-            constexpr uint32_t LSR_TX_READY_POS = 5U;
-            return this->regs.m_lsr | (1U << LSR_TX_READY_POS);//We are always ready to transmit
+            //No need to implement any of the error bits, just the TX and RX ready bits
+            constexpr uint32_t LSR_TX_NOT_IN_PROGRESS_POS   = 6U;
+            constexpr uint32_t LSR_TX_READY_POS             = 5U;
+            constexpr uint32_t LSR_RX_READY_POS             = 0U;
+            uint8_t lsr = 0;
+            lsr |= 1U << LSR_TX_NOT_IN_PROGRESS_POS;//We are always ready to transmit
+            lsr |= 1U << LSR_TX_READY_POS;//We are always ready to transmit
+            lsr |= (receive_queue.size() > 0) ? (1U << LSR_RX_READY_POS) : 0;//Set this if characters are waiting!
```

Comment:
First part of comment from 135 also applies here

Reply:
Will do!